File: core/kora-app-annotation-processor/src/test/java/io/koraframework/kora/app/annotation/processor/app/AppWithFactories13.java
package io.koraframework.kora.app.annotation.processor.app;

import io.koraframework.application.graph.TypeRef;
import io.koraframework.common.annotation.KoraApp;
import io.koraframework.common.annotation.Root;

@KoraApp
public interface AppWithFactories13 {

    @Root
    default Object target(ConfigValueMapper<TestEnum> object) {
        return new Object();
    }

    default <T extends Enum<T> & MarkerOne> ConfigValueMapper<T> extractor1(TypeRef<T> ref) {
        return value -> null;
    }

    default <T extends Enum<T> & MarkerTwo> ConfigValueMapper<T> extractor2(TypeRef<T> ref) {
        return value -> null;
    }

    enum TestEnum implements MarkerOne {}

    interface MarkerOne { }

    interface MarkerTwo { }

    interface ConfigValueMapper<T> {

        T extract(Object value);
    }
}

File: core/kora-app-symbol-processor/src/test/kotlin/io/koraframework/kora/app/ksp/app/AppWithFactories13.kt
package io.koraframework.kora.app.ksp.app

import io.koraframework.application.graph.TypeRef
import io.koraframework.common.annotation.KoraApp
import io.koraframework.common.annotation.Root

@KoraApp
interface AppWithFactories13 {

    @Root
    fun target(cl: ConfigValueMapper<TestEnum>): Any = Any()

    fun <T> extractor1(ref: TypeRef<T>): ConfigValueMapper<T> where T : Enum<T>, T : MarkerOne {
        return object : ConfigValueMapper<T> {
            override fun extract(value: Any?): T? = null
        }
    }

    fun <T> extractor2(ref: TypeRef<T>): ConfigValueMapper<T> where T : Enum<T>, T : MarkerTwo {
        return object : ConfigValueMapper<T> {
            override fun extract(value: Any?): T? = null
        }
    }

    enum class TestEnum : MarkerOne

    interface MarkerOne

    interface MarkerTwo

    interface ConfigValueMapper<T> {

        fun extract(value: Any?): T?
    }
}
